session/database.rst

```diff
@@ -0,0 +1,600 @@
+.. index::
+    single: Session; Database Storage
+
+Store Sessions in a Database
+============================
+
+Symfony stores sessions in files by default. If your application is served by
+multiple servers, you'll need to use instead a database to make sessions work
+across different servers.
+
+Symfony can store sessions in all kinds of databases (relational, NoSQL and
+key-value) but recommends key-value databases like Redis to get best performance.
+
+Store Sessions in a key-value Database (Redis)
```

Comment:
I'm quite sure that the RedisSessionHandler does not perform session locking (unlike the default file storage, or the PDO session handler), which then mean that you can have race conditions on session access

Such race condition happened to me in the past when using the (now removed) DbalSessionHandler which was not locking session: a request would change the session to store a CSRF token in it. But due to a concurrent request also using the session in parallel and saving it after (so without the CSRF token in it as it was not there when the session was loaded), the following form submission would get an "Invalid CSRF token" error (because the session write done when generating the CSRF would be lost due to the race condition).
I spent weeks of seeing this error regularly before understanding that it was caused by the lack of locking in the handler.

Note that locking can cause slowdowns if you have multiple concurrent requests from the same session, for instance in AJAX-heavy interfaces, as requests have to wait to acquire the lock. So that's a choice to make (you could decide to favor availability over consistency for your session storage),. But the choice needs to be done based on awareness.
And this is the reason why locking is optional in PdoSessionHandler (it is turned on by default, but there is a `LOCK_NONE` lock mode).